### ISSUE TITLE
LameStation support

### ISSUE THREAD

**Issue by DarkmatterVale** [parallaxinc/BlocklyProp#40]
LameStation uses Spin on the Propeller to compile, but a separate set of libraries.

**Comment by michel-cf** [comment]
If it uses spin objects, all that needs to be done is adding the used spin files in the correct directory of the BlocklyPropClient project and you can start using them in Blocks

**Comment by DarkmatterVale** [comment]
True, but it would be preferable to have it as a separate generator because it doesn't really fit with the Spin generator. It would still use Spin, but the interface ( what blocks the user can add ) should be a little different. Most of the traditional Spin blocks should not be available to the device since a lot of the functions are already integrated inside Mr. Weir's libraries.

**Comment by michel-cf** [comment]
Just create another Blockly version (outer page and frame, and load a different set of Blocks.

You'll be able to reuse some while leaving out others. (By choosing what js files to include and by listing  the block names. This might possibly change with the Blockly upgrade)

**Comment by DarkmatterVale** [comment]
Oh, that makes a lot more sense! I will do that as soon as I know what files I need to include to add support :).

How will the upgrade affect this?

**Comment by PropGit** [comment]
Michel,

Will you plan on describing this and possibly demonstrating it when you visit us?

**Comment by DarkmatterVale** [comment]
@PropGit,

The upgrade or LameStation support?

**Comment by michel-cf** [comment]
If with this, you mean the LameStation support. I sure can explain what needs to be done to add the functionality.

What blocks need to be added needs to be reviewed. I once watched the demo but never really studied the necessary code. Demonstrating will be an even bigger problem, as I don't have access to a LameStation and the Blockly upgrade is taking a lot of my little available time. Maybe Vale can create a minimal version that we could test when I'm visiting.

I think we'll have to create higher level blocks then up till now, that combine multiple actions in one. As stated before, this depends on how code needs to be written with the LameStation platform.

**Comment by DarkmatterVale** [comment]
I am currently working with Mr. Weir on what code needs to be written to add support. It uses Spin, so in depth changes should be minimal if any. In terms of a demo, I should be able to put something together by then. It all depends how much work is required to add support.

I will report back with an idea of what blocks need to be added once I can get a hold of Mr. Weir.

**Comment by DarkmatterVale** [comment]
I have communicated with Mr. Weir, and below is the summery of my conversation with him.

1) Content tools need to be added. These include img2dat+LSPaint and eventually wav2dat ( I do not know the status of this content tool ). I think these will need to be added in the Client, and run from there.
2) Dynamic block support. Basically, we need to be able to generate blocks from spin files. I do not yet know if this is possible, Michel do you know?
3) To program, you use a set of libraries in Spin and compile on any supported compiler ( OpenSpin will work )

**Comment by DarkmatterVale** [comment]
During my talk with Mr. Weir, he made a very good point about maintaining our blocks. Since code will update, it will be very difficult for us to keep our blocks maintained with the newest version of code. So in light of this, I am working on creating a tool that will convert Spin files ( v1.0 will just be Spin ) to .js files that can be copied right into the spin generator folder. This will make maintaining our blocks VERY easy. I will post more info when I have more if it completed.

@PropGit:

So, if you were talking about LameStation support no, I will not have something ready by the time Michel comes out to visit you guys. I can have something done by the end of April ( at the latest hopefully ).

**Comment by michel-cf** [comment]
Creating a system that creates blocks on the fly based on spin code files has always been on the idea list. It would be great if you could could add it.

My idea essentially was that the user can select a spin file through BlocklyProp. The code gets analyzed (either in the browser or on the server). Blocks for it get created and stored in the database. Together with the spin code file. If someone wants to compile a program that uses custom blocks, it checks if their BlocklyPropClient has the custom spin files and if not, BlocklyProp will download them from the server and upload them to the BlocklyPropClient, where it gets added to the spin library.

**Comment by DarkmatterVale** [comment]
Michel,

I am not planning on adding that ability for a user to implement custom Spin blocks right now. I am talking about a dedicated program for us to use when updating blocks. This will make my life a lot easier with library updates etc.

That would be a great addition though!

**Comment by DarkmatterVale** [comment]
Since my converter is done, Mr. Weir now needs to implement the changes and convert all of the lamentation functions into blocks. Once this is done, we should have support for LameStation on the graphical side of things!

**Comment by DarkmatterVale** [comment]
This is not a priority at the moment and can be reopened later if necessary.
